protobuf-based rasctl/rascsi command interface

uweseimet opened the pull request:
This is just a draft on how a protobuf-based approach for an interface to return results from rascsi to rasctl might look like. Instead of using fputs the binary protobuf data would have to be returned and rasctl would read the protobuf result similar to how rascsi writes it to the protocol buffer. The final data structure would contain not just the string but at least a status value on which rasctl can react.
@akuker Before spending more time on this I'd like to know if you are OK with this approach as a resolution for issue #43.

akuker commented:
For the auto-generated files, what is the input to define the structures?

I like what's there so far!

uweseimet commented:
The ..proto file (which I checked in) is the input. See https://developers.google.com/protocol-buffers/docs/cpptutorial for examples.

uweseimet commented:
Just a not for the record: protobuf for c++ has to be installed on the Pi with:

sudo apt install c++-protobuf protobuf-compiler

uweseimet commented:
The rascsi process now returns structured data (using protobuf) and rasctl deserializes these data, so that the output looks like before:

```
rascsi>./bin/fullspec/rasctl  -i 3 -c a -f ~/images/scsi3.hds
rascsi>./bin/fullspec/rasctl  -i 3 -c i -f ~/images/scsi3.hds
Error : Operation denied (Device type SCHD isn't removable)
```
The data returned is not a simple string anymore but a string and a status code, see the .proto file for details

@akuker I don't think the baremetal build can profit from this change, and I added an ifdef so that for baremetal a simple string is returned, just like before. I cannot test this, though, I guess you can?
I wonder how important the baremetal build actually is. Are they any advantages of not running rascsi on an OS? I expect the baremetal environment to be not very convenient.

akuker commented on src/raspberrypi/scsi.h:
```diff
@@ -12,6 +12,28 @@
+class ERROR_CODES
+{
+public:
+	enum sense_key : int {
```
Would it be possible to fully define these enums with all of the options in the spec?

uweseimet replied:
The spec contains really a lot codes (ASCs), because it covers so many devices. This is why I'd suggest to only define those needed for RaSCSI. Or to make this a separate issue, that may be addressed in the future.
In case you are only referring to the sense key, and not to the ASCs I can simply do this as part of this PR.

akuker replied:
Just the sense keys is fine.

uweseimet replied:
Done

akuker commented:
> @akuker I don't think the baremetal build can profit from this change, and I added an ifdef so that for baremetal a simple string is returned, just like before. I cannot test this, though, I guess you can?
> I wonder how important the baremetal build actually is. Are they any advantages of not running rascsi on an OS? I expect the baremetal environment to be not very convenient.

@uweseimet - I'm pretty sure that I've broken the baremetal build through several of my changes. I haven't been testing it. I've had a few folks who want to use it, and I've pointed them to the original GIMONS version. I agree that the functionality is much more limited. 

The desire with the baremetal version is the fast startup time for instances where the RaSCSI is mounted inside a PC (and gets power from the PC). With a default Raspberry Pi OS, it can take up to 30 seconds to boot up. Depending upon the configuration, the host might time out waiting  for the RaSCSI to boot.

Personally, I'd prefer to focus on getting a stripped down Buildroot/Yocto image that can boot within a few seconds rather than going with an "instant on" bare metal. (And I've been slowly working on that in the background, with my limited free time)

uweseimet commented:
@akuker Thank for commenting on the baremetal build.

Regarding this PR I think it makes sense to use protobuf for both communication channels, so that code like

```
	sprintf(buf, "%d %d %d %d %s\n", id, un, cmd, type, file ? file : "-");
	if (!SendCommand(buf)) {
```
can be replaced by a data structure. Any future extensions can then easily (and typesafe) be implemented when needed.

akuker commented:
I whole-heartedly agree!

uweseimet commented:
The communication now uses protobuf in both directions and uses protobuf enums in the protobuf interface. These enums are more intelligent than C++ enums, in particular they can return their stringified names, so that I could do some more code cleanup.

uweseimet commented:
@akuker This feature branch is ready for review and merge. Regarding the merge conflicts, just take the code from this branch and discard the old code in the develop branch.

I added rasutil.cpp with convenience methods for serializing/deserializing protobuf data and also cleaned up and simplified some parts of the code.
Testing with my retro hardware did not reveal any issues.

uweseimet commented:
@akuker I just added an option -s (-l is already used) to rasctl, which changes the log level at runtime. This addresses https://github.com/akuker/RASCSI/issues/19.

akuker commented on src/raspberrypi/rasutil.cpp:
```diff
@@ -0,0 +1,69 @@
+}
+
+string DeserializeProtobufData(int fd)
+{
```
Would you mind adding a few comments throughout this function?

uweseimet replied:
Done

akuker commented on src/raspberrypi/rasctl.cpp:
```diff
@@ -117,158 +115,142 @@ int main(int argc, char* argv[])
-			exit(EINVAL);
-		}
+	// File check (command is ATTACH and type is HD, CD or MO)
+	if (cmd == ATTACH && (type == SASI_HD || type == SCSI_HD || type == MO || type == CD) && params.empty()) {
```
Why can't we attach a MO/CD device without a file path? This would just create a drive without a disk inserted.

uweseimet replied:
Thank you for spotting this. I did not realize that not having a path is a valid use case.

uweseimet commented:
@akuker Please wait for my go before merging. I just stumbled upon an issue with returning the error status, which I would like to investigate.

uweseimet commented:
@akuker The problem is resolved, ready for merge. This PR also resolves https://github.com/akuker/RASCSI/issues/43:

>rasctl -i 3 -c attach -t hd -f /home/pi/images/test.hda ; echo $?
Error : File open error [/home/pi/images/test.hda]
107

Please confirm that https://github.com/akuker/RASCSI/issues/43 can be closed.

akuker commented:
@uweseimet - could you update easyinstall.sh in the root of the repo to add these on line 44?
libprotobuf17 protobuf-compiler

akuker commented on src/raspberrypi/rascsi.cpp:
```diff
@@ -451,130 +444,173 @@ void MapControler(FILE *fp, Disk **map)
+	string params = command.params().c_str();
+
+	ostringstream s;
+	s << "Processing: cmd=" << Operation_Name(cmd) << ", id=" << id << ", un=" << un
```
@uweseimet 

```
rascsi.cpp: In function ‘bool ProcessCmd(FILE*, const rasctl_interface::Command&)’:
rascsi.cpp:525:29: error: ‘Operation_Name’ was not declared in this scope
rascsi.cpp:526:20: error: ‘DeviceType_Name’ was not declared in this scope

```

akuker replied:
Did you forget to check in some updates?

uweseimet replied:
I don't think something is missing, at least I don't get this error. Try "make clean" before recompiling. Note that I'm currently compiling under Ubuntu. If you still have issues after "make clean" I will boot Raspian OS and verify.

uweseimet replied:
Just tried Raspian OS, and now I can reproduce your issue. Working on it ...

uweseimet replied:
@akuker Too bad, the outdated Raspbian protoc does not support converting enums to their names. I reverted parts of the code to not do this conversion. On my Raspbian system everything compiles now, please verify.

uweseimet commented:
> @uweseimet - could you update easyinstall.sh in the root of the repo to add these on line 44?
> libprotobuf17 protobuf-compiler

Are you sure that this should not be "c++-protobuf protobuf-compiler" as mentioned in one of my previous comments?

uweseimet commented:
> > @uweseimet - could you update easyinstall.sh in the root of the repo to add these on line 44?
> > libprotobuf17 protobuf-compiler
> 
> Are you sure that this should not be "c++-protobuf protobuf-compiler" as mentioned in one of my previous comments?

@akuker I checked this and found that only "protobuf-compiler" is needed, at least when removing c++-protobuf everything still worked for me. I updated easyinstall.sh accordingly, please verify.
